src/main/java/io/camunda/testing/filters/ProcessInstanceRecordStreamFilter.java

```diff
@@ -0,0 +1,58 @@
+package io.camunda.testing.filters;
+
+import io.camunda.zeebe.protocol.record.Record;
+import io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent;
+import io.camunda.zeebe.protocol.record.value.BpmnElementType;
+import io.camunda.zeebe.protocol.record.value.ProcessInstanceRecordValue;
+import java.util.Arrays;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class ProcessInstanceRecordStreamFilter {
+
+  private Stream<Record<ProcessInstanceRecordValue>> stream;
+
+  public ProcessInstanceRecordStreamFilter(
+      final Iterable<Record<ProcessInstanceRecordValue>> records) {
+    stream = StreamSupport.stream(records.spliterator(), false);
+  }
+
+  public ProcessInstanceRecordStreamFilter withProcessInstanceKey(final long processInstanceKey) {
+    stream =
+        stream.filter(record -> record.getValue().getProcessInstanceKey() == processInstanceKey);
+    return this;
```

Comment:
:wrench: We could return a new object here, immutable function programming style. Same in the other methods